packages/roc-package-web-app-react/app/server/render.js

```diff
@@ -99,13 +96,11 @@ export function reactRender(url, createRoutes, store, renderPage, staticRender =
                     };
                 }
 
-                getPrefetchedData(components, locals)
+                trigger('fetch', components, locals)
                     .then(() => {
-                        let component = <RoutingContext {...renderProps} />;
+                        let component = <RouterContext {...renderProps} />;
```

Comment:
I see, now it makes sense! 🙂

Most of this is already available through the current structure, just a bit different in the APIs and how Redux is used. Instead of managing transition with Redux on the client side we are using a `RouterContext`. This is since version 1.0 of React Router the recommended way to do this sort of things. Currently it is coded to load `fetch` and `defer` simultaneously but make the transition as soon as the `fetch` is completed __if__ `blocking` in settings is defined to `true` (default). _I see that this might not be what one wants, it might be more logical to first completely load `fetch` and then `defer` if `blocking` is `true`.

The server will, currently, only prefetch `fetch` as you have already mentioned. The reasoning for this is that we might want to render the page as soon as possible and then being able to _defer_ some data fetching to the client can be beneficial. I however agree with you that in some cases this is not the expected/wanted behaviour and I suggest the following.

A setting that defines what Redial _"names"_ that should be fetched on the server before rendering. 

```js
// Not the actual syntax
settings.runtime.fetchOnServer = ['fetch', 'defer', 'only-server']
```  
Then we probably want to do something similar on the client but for before and after route transition, if `blocking` is `true`.
```js
// Not the actual syntax
settings.runtime.fetchOnClient = {
  blocking: ['fetch'],
  nonBlocking: ['defer', 'client-only']
}
```
----
Since we actually can create anything it might be interesting to be able to define something similar to lifecycle things. An example of this would be `done` that maybe should run after all the other "blocking" things have been completed. This would allow these functions to have an updated state from the previous ones.